### PR TITLE
Combine brine + co2store

### PR DESCRIPTION
**Pull request by totto82**
This allows for combination of CO2STORE and BRINE. 

Need some more testing before it can be merge i.e. lets make it a draft PR.

### CODE REVIEW

File: src/opm/material/fluidsystems/BlackOilFluidSystem.cpp [hunk 1]
```diff
@@ -118,11 +118,15 @@ initFromState(const EclipseState& eclState, const Schedule& schedule)
...
+        const Scalar MmNaCl = 58e-3; // molar mass of NaCl [kg/mol]
+        // convert to mass fraction
+        const Scalar salinity = 1 / ( 1 + 1 / (molality*MmNaCl));
         for (unsigned regionIdx = 0; regionIdx < numRegions; ++regionIdx) {
             if (phaseIsActive(oilPhaseIdx)) // The oil component is used for the brine if OIL is active
-                molarMass_[regionIdx][oilCompIdx] = BrineCo2Pvt<Scalar>::Brine::molarMass();
+                molarMass_[regionIdx][oilCompIdx] = BrineCo2Pvt<Scalar>::Brine::molarMass(salinity);
             if (phaseIsActive(waterPhaseIdx))
-                molarMass_[regionIdx][oilCompIdx] = BrineCo2Pvt<Scalar>::Brine::molarMass();
+                molarMass_[regionIdx][oilCompIdx] = BrineCo2Pvt<Scalar>::Brine::molarMass(salinity);
```
Comment (goncalvesmachadoc) [inline src/opm/material/fluidsystems/BlackOilFluidSystem.cpp:129]: Both if's here are doing exactly the same.

Comment (totto82) [reply src/opm/material/fluidsystems/BlackOilFluidSystem.cpp]: Thanks for noticing. This is now fixed in the master branch in a separate PR

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
I have done some initial testing and the PR is thus ready for review.

File: opm/material/components/BrineDynamic.hpp [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+     * \copydoc Component::molarMass
+     *
+     * This assumes that the salt is pure NaCl.
+     */
+    template <class Evaluation>
+    static Evaluation molarMass(const Evaluation& salinity)
+    {
+        const Scalar M1 = H2O::molarMass();
+        constexpr Scalar M2 = 58e-3; // molar mass of NaCl [kg/mol]
+        const Evaluation X2 = salinity; // mass fraction of salt in brine
```
Comment (goncalvesmachadoc) [inline opm/material/components/BrineDynamic.hpp:83]: Here you have a different value from  Brine_CO2.hpp (`const Scalar Ms = 58.8e-3; /* molecular weight of NaCl  [kg/mol] */`) . Shouldn't be the same value for molar mass and molar weight?

File: src/opm/material/fluidsystems/BlackOilFluidSystem.cpp [hunk 3]
```diff
@@ -118,11 +118,15 @@ initFromState(const EclipseState& eclState, const Schedule& schedule)
     // use molarMass of CO2 and Brine as default
     // when we are using the the CO2STORE option
     if (eclState.runspec().co2Storage()) {
+        const Scalar molality = eclState.getTableManager().salinity(); // mol/kg
+        const Scalar MmNaCl = 58e-3; // molar mass of NaCl [kg/mol]
+        // convert to mass fraction
```
Comment (goncalvesmachadoc) [inline src/opm/material/fluidsystems/BlackOilFluidSystem.cpp:123]: Here again.

File: opm/material/binarycoefficients/Brine_CO2.hpp [hunk 4]
```diff
@@ -212,14 +212,15 @@ class Brine_CO2 {
      *
      * \param salinity the salinity [kg NaCl / kg solution]
      */
-    static Scalar salinityToMolFrac_(Scalar salinity) {
+    template <class Evaluation>
+    static Evaluation salinityToMolFrac_(const Evaluation& salinity) {
 
         const Scalar Mw = H2O::molarMass(); /* molecular weight of water [kg/mol] */
         const Scalar Ms = 58.8e-3; /* molecular weight of NaCl  [kg/mol] */
```
[the remark this answers is not in the document]
Comment (totto82) [reply opm/material/binarycoefficients/Brine_CO2.hpp:219]: Thanks for noticing. I will unify and correction the molecular weight of NaCl here and elsewhere in the code.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
The change of molar mass of NaCl changes the results and the test failures are thus expected.

**Comment by totto82** [comment]
Any more comments? Or can we merge this after  we have updated the ref-data.

**Comment by goncalvesmachadoc** [comment]
No more comments on my side. LGTM.

**Review by bska — COMMENTED** [review]
Is there a plan to add some unit tests for this facility?  There are a lot of hard-coded numbers in this work, and I'd really like it if there were tests in place to make probable that we calculate the expected numbers.

**Comment by totto82** [comment]
> Is there a plan to add some unit tests for this facility? There are a lot of hard-coded numbers in this work, and I'd really like it if there were tests in place to make probable that we calculate the expected numbers.

I see your point. Note that the brine properties are the same as in Brine.hpp, only the interface is changed, but I have kept the old Brine.hpp for compatibility. Adding a unit test that checks that Brine.hpp and BrineDynamic.hpp computes the same values is probably a good idea. That would add a small guard against fixing issues in one of files and not in the other.

**Comment by bska** [comment]
> Adding a unit test that checks that Brine.hpp and BrineDynamic.hpp computes the same values is probably a good idea.

Agreed, but I was also thinking something along the lines of "in these conditions the brine density should be this (known/pre-computed) value".

**Comment by totto82** [comment]
> Agreed, but I was also thinking something along the lines of "in these conditions the brine density should be this (known/pre-computed) value".

Again, I see your point. But I will argue that such a test shouldn't stop this PR, since this PR does not change the current implementation of the Brine class.

**Comment by totto82** [comment]
jenkins build this please

**Comment by bska** [comment]
> But I will argue that such a test shouldn't stop this PR, since this PR does not change the current implementation of the Brine class.

That's why I asked if there was a **plan** to add such tests.  The code imported from what used to be opm-material is not sufficiently well covered by unit tests and that makes working on it very hard for those who are not very well acquainted with its internals.  Going forward I may not be as willing to merge these kinds of changes without associate unit tests.

**Comment by totto82** [comment]
> That's why I asked if there was a plan to add such tests. 

Thanks for pointing this out. I agree that this is an urgent issue that should be prioritized.

**Comment by totto82** [comment]
jenkins build this opm-tests=964 please

**Comment by totto82** [comment]
jenkins build this opm-tests=964 please

**Comment by bska** [comment]
jenkins build this opm-tests=964 please

**Pull request merged**